drivers/wifi/esp8266/esp8266.c

```diff
@@ -0,0 +1,844 @@
+			     ESP_CMD_TIMEOUT);
+	if (ret < 0) {
+		LOG_ERR("Failed to enable AP mode, ret %d", ret);
+		return ret;
+	}
+
+	len = snprintk(cmd, sizeof(cmd), "AT+"_CWSAP"=\"");
+	memcpy(&cmd[len], params->ssid, params->ssid_length);
+	len += params->ssid_length;
+
+	if (params->security == WIFI_SECURITY_TYPE_PSK) {
+		len += snprintk(&cmd[len], sizeof(cmd) - len, "\",\"");
+		memcpy(&cmd[len], params->psk, params->psk_length);
+		len += params->psk_length;
+		ecn = 3;
+	} else {
+		len += snprintk(&cmd[len], sizeof(cmd) - len, "\",\"");
+	}
+
+	snprintk(&cmd[len], sizeof(cmd) - len, "\",%d,%d", params->channel,
+		 ecn);
+
+	ret = modem_cmd_send(&data->mctx.iface, &data->mctx.cmd_handler,
+			     NULL, 0, cmd, &data->sem_response,
+			     ESP_CMD_TIMEOUT);
+
+	return ret;
+}
+
+static int esp8266_mgmt_ap_disable(struct device *dev)
+{
+	struct esp8266_data *data = dev->driver_data;
+	int ret;
+
+	ret = modem_cmd_send(&data->mctx.iface, &data->mctx.cmd_handler,
+			     NULL, 0, "AT+"_CWMODE"=1", &data->sem_response,
+			     ESP_CMD_TIMEOUT);
+
+	return ret;
+}
+
+static void esp8266_init_work(struct k_work *work)
+{
+	struct esp8266_data *dev;
+	int ret;
+	static struct setup_cmd setup_cmds[] = {
+		/* turn off echo */
+		SETUP_CMD_NOHANDLE("ATE0"),
+#if defined(CONFIG_WIFI_ESP8266_AT_VERSION_2_0)
+		SETUP_CMD_NOHANDLE("AT+CWAUTOCONN=0"),
+#endif
+		SETUP_CMD_NOHANDLE("AT+UART_CUR=115200,8,1,0,0"),
+		/* enable multiple socket support */
+		SETUP_CMD_NOHANDLE("AT+CIPMUX=1"),
+		SETUP_CMD_NOHANDLE("AT+"_CWMODE"=1"),
+		/* only need ecn,ssid,rssi,channel */
+		SETUP_CMD_NOHANDLE("AT+CWLAPOPT=0,23"),
+#if defined(CONFIG_WIFI_ESP8266_PASSIVE_TCP)
+		SETUP_CMD_NOHANDLE("AT+CIPRECVMODE=1"),
+#endif
+		SETUP_CMD("AT+"_CIPSTAMAC"?", "+"_CIPSTAMAC":",
+			  on_cmd_cipstamac, 1U, ""),
+	};
+
+	dev = CONTAINER_OF(work, struct esp8266_data, init_work);
+
+	ret = modem_cmd_handler_setup_cmds(&dev->mctx.iface,
+					   &dev->mctx.cmd_handler, setup_cmds,
+					   ARRAY_SIZE(setup_cmds),
+					   &dev->sem_response,
+					   ESP_INIT_TIMEOUT);
+	if (ret < 0) {
+		LOG_ERR("Init failed %d", ret);
+		return;
+	}
+
+	net_if_set_link_addr(dev->net_iface, dev->mac_addr,
+			     sizeof(dev->mac_addr), NET_LINK_ETHERNET);
+
+	LOG_INF("ESP8266 Wi-Fi ready");
+
+	net_if_up(dev->net_iface);
+
+	k_sem_give(&dev->sem_if_up);
+}
+
+static void esp8266_reset(struct esp8266_data *dev)
+{
+	int ret;
+
+	if (net_if_is_up(dev->net_iface)) {
+		net_if_down(dev->net_iface);
+	}
+
+#if defined(DT_INST_0_ESPRESSIF_ESP8266_WIFI_RESET_GPIOS_PIN)
+	modem_pin_write(&dev->mctx, WIFI_RESET, 0);
+	k_sleep(K_MSEC(100));
+	modem_pin_write(&dev->mctx, WIFI_RESET, 1);
+#else
+	/*
+	 * FIXME:
+	 * Need to wait for esp8266 to boot. Better to do some resends
+	 * instead of a sleep.
+	 */
+	k_sleep(200);
+
+	ret = modem_cmd_send(&dev->mctx.iface, &dev->mctx.cmd_handler,
+			     NULL, 0, "AT+RST", &dev->sem_response,
+			     ESP_CMD_TIMEOUT);
```

Comment:
This will cause some issues as you are using a handler that is looking for "ready" and AT+RST is only a soft reset which returns OK.
It could be replaced with AT+RESTORE which forces a full reset but you will lose any stored configuration, but does return the firmware "ready" you are looking for.

Additionally, the esp32 UART boots at 921600 not the 115200 that the AT firmware uses, so will struggle to see the firmware ready response.

Reply:
You can now select UART speed and flow-control with the `wifi-uart-speed` and `wifi-uart-flow-control` in the device tree. When https://github.com/zephyrproject-rtos/zephyr/pull/23245 gets merged I think those can be removed and speed and flow-control can be taken directly from the UART bus node.

Regarding the `ready`, are you sure that doesn't appear after a `AT+RST`? On the esp8266, a `AT+RST` is followed by `OK` but after it restarts I also get the `ready`.

The reason for acting on the `ready` keyword is actually to handle the cases where the module resets which I've had happened to me sometimes.